conn/sticky_linux.go

```diff
@@ -0,0 +1,121 @@
+/* SPDX-License-Identifier: MIT
+ *
+ * Copyright (C) 2017-2022 WireGuard LLC. All Rights Reserved.
+ */
+
+package conn
+
+import (
+	"net/netip"
+	"unsafe"
+
+	"golang.org/x/sys/unix"
+)
+
+// TODO: macOS, FreeBSD and other BSDs likely do support this feature set, but
+// use alternatively named flags and need ports and require testing.
+
+// getSrcFromControl parses the control for PKTINFO and if found updates ep with
+// the source information found.
+func getSrcFromControl(control []byte, ep *StdNetEndpoint) {
+	ep.ClearSrc()
+
+	var (
+		hdr  unix.Cmsghdr
+		data []byte
+		rem  []byte = control
+		err  error
+	)
+
+	for len(rem) > unix.SizeofCmsghdr {
+		hdr, data, rem, err = unix.ParseOneSocketControlMessage(control)
+		if err != nil {
+			return
+		}
+
+		if hdr.Level == unix.IPPROTO_IP &&
+			hdr.Type == unix.IP_PKTINFO {
+
+			info := bufCast[unix.Inet4Pktinfo](data)
+			ep.src.Addr = netip.AddrFrom4(info.Spec_dst)
+			ep.src.ifidx = info.Ifindex
+
+			return
+		}
+
+		if hdr.Level == unix.IPPROTO_IPV6 &&
+			hdr.Type == unix.IPV6_PKTINFO {
+
+			info := bufCast[unix.Inet6Pktinfo](data)
+			ep.src.Addr = netip.AddrFrom16(info.Addr)
+			ep.src.ifidx = int32(info.Ifindex)
+
+			return
+		}
+	}
+}
+
+// cast a buffer to a type directly, only for types known to be safe to do so.
+// panics if buf is of insufficient size.
+func bufCast[T unix.Inet4Pktinfo | unix.Inet6Pktinfo](buf []byte) (t T) {
+	if len(buf) < int(unsafe.Sizeof(t)) {
+		panic("byteCast: buffer too small")
+	}
+	return *(*T)(unsafe.Pointer(&buf[0]))
```

Comment:
This is a cool macro. I started to wonder, "why isn't there one of these in std that works over all structs generically? what would be unsafe?" Answer: alignment. Except these pktinfo structs have int32s in them, so directly casting an arbitrary address will break alignment (on platforms nobody cares about). However, it looks like you're returning a copy anyway, so maybe it doesn't matter? I guess it depends on how Go does that copy. The safer way is to just allocate and memcpy, though, and rely on the compiler doing word-sized loads and stores on x86 and byte-sized ones on arm32. (In which case, again, why doesn't std's unsafe package have a nice helper for all generic structs?) ORRR maybe you know something about the alignment of &buf[0], in which case you should document that or enforce it or something.

Reply:
I updated this to use `copy()` in https://github.com/WireGuard/wireguard-go/pull/64/commits/0871fd456922413c01eef744e24e69dcaa39cbd2.

Reply:
`func (v *virtioNetHdr) decode(b []byte) error` checks the length of the buffer (and panics) before doing the copy but this new one doesn't. Up to you to decide whether to add to missing one, remove from prior one, or keep as-is for good reasons. I just noticed the difference so thought I should point it out.

Reply:
Btw any thoughts on something like this in std?

```go
func unsafe.MarshalStructFrom[T any](t *T, buf []byte) {
	l := unsafe.Sizeof(*t)
	if l > len(buf) { panic(...) }
	copy(unsafe.Slice((*byte)(unsafe.Pointer(t)), l), buf)
}
func unsafe.MarshalStructTo[T any](buf []byte, t *T) {
	l := unsafe.Sizeof(*t)
	if l > len(buf) { panic(...) }
	copy(buf[:l], unsafe.Slice((*byte)(unsafe.Pointer(t)), l))
}
```

Just writing on my phone. But maybe interesting? Cc @bradfitz

Reply:
I added a panic'ing len check in `pktInfoFromBuf()` in https://github.com/WireGuard/wireguard-go/pull/64/commits/9536cec4c50eae7b72575117f3f8561e39ef730d.

I'll let others chime in on the std idea as I'm not familiar with contributing to unsafe.